setup/helpers/abs_helper.rb

```diff
@@ -635,43 +641,156 @@ def verify_abs_hosts(hosts)
   end
 
-  # Determines whether the specified hosts list is valid
+  # Determines whether the specified JSON string can be parsed
+  # Validates the parsed output to ensure the first element contains a 'hostname' key
   #
   # @author Bill Claytor
   #
-  # @param [JSON] abs_resource_hosts The hosts to validate
+  # @param [JSON] abs_resource_hosts The JSON string to validate
   #
-  # @return [true, false] Based on the validity of the specified hosts
+  # @return [true, false] Based on the validity of the specified JSON string
   #
   # @example
-  #   valid = abs_valid_resource_hosts?(abs_resource_hosts)
+  #   json_string = "[{"hostname":"ip-10-227-2-105.amz-dev.puppet.net","type":"centos-7-x86_64","engine":"aws"}]"
+  #   valid = abs_valid_resource_hosts?(json_string)
+  #   valid == true
+  #
+  #   json_string = "[{"abc":"123","xyz":"567"}]"
+  #   valid = abs_valid_resource_hosts?(json_string)
+  #   valid == false
   #
   def valid_abs_resource_hosts?(abs_resource_hosts)
     is_valid = false
 
+    # initially included this here since it is validating the string being read from a file
+    # TODO: move this check into parse_abs_resource_hosts as well?
     if abs_resource_hosts.nil?
-      puts "A valid hosts array is required; nil was specified"
+      puts "A valid JSON string is required; nil was specified"
       puts
     else
 
       begin
-        hosts = JSON.parse(abs_resource_hosts)
-        host = hosts[0]
-        hostname = host["hostname"]
-        if !hostname.nil? && !hostname.empty?
-          is_valid = true
-        else
-          puts "The specified resource host array is not valid: #{abs_resource_hosts}"
-          puts
-        end
-      rescue StandardError
+        parse_abs_resource_hosts(abs_resource_hosts)
+        is_valid = true
+      rescue StandardError => e
         # TODO: raise?
-        puts "JSON::ParserError encountered parsing the hosts array: #{abs_resource_hosts}"
+        puts e.message.to_s
       end
-
     end
 
     is_valid
   end
-  # rubocop:enable Naming/AccessorMethodName
+
+  # Parses the specified JSON string
+  # Validates the parsed output to ensure the first element contains a 'hostname' key
+  # Returns the parsed hosts array
+  #
+  # note: consolidating this here since it is used by multiple methods
+  #
+  # @author Bill Claytor
+  #
+  # @param [JSON] abs_resource_hosts The hosts JSON to parse
+  #
+  # @raise [RuntimeError] If parsing is not successful
+  #
+  # @return [Array] The hosts array
+  #
+  # @example
+  #   json_string = "[{"hostname":"ip-10-227-2-105.amz-dev.puppet.net","type":"centos-7-x86_64","engine":"aws"}]"
+  #   hosts = parse_abs_resource_hosts(json_string)
+  #
+  def parse_abs_resource_hosts(abs_resource_hosts)
+    puts "Parsing the specified abs_resource_hosts JSON:"
+    puts abs_resource_hosts
+    puts
+
+    begin
+      hosts = JSON.parse(abs_resource_hosts)
+      hostname = hosts[0]["hostname"]
+      if hostname.nil? && hostname.empty?
```

Comment:
:bug: this should be an OR not an AND condition.  I discovered this with the following modified spec test.
```
    context "when an invalid hosts JSON string is specified" do
      invalid_hosts = [
        nil,                      # nil case
        'Not valid JSON',         # not JSON
        '{}',                     # valid JSON, not hosts array
        '[{"hostname": ""}]',     # hostname.is_empty?
        '[{ "hostname": "legit" }, {} ]'  # invalid host[1]
      ]
      invalid_hosts.each do |i|
        it "raises an error" do
          expect { subject.parse_abs_resource_hosts(i) }
            .to raise_error(RuntimeError, /encountered parsing the hosts array/)
        end
      end
    end
```

Reply:
Updated to add tests for invalid key and empty value.